typing/typecore.ml

```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
+      in
+      type_approx env e ty_expected
+  | Pexp_coerce (_, sty1, sty2) ->
+      ignore @@
+      type_approx_constraint env (Pcoerce (sty1, sty2)) ~loc ty_expected
   | Pexp_pack (_, Some ptyp) ->
-      let ty = newvar () in
       let sty = Ast_helper.Typ.package ~loc ptyp in
-      type_approx_constraint env ty (Pconstraint sty) ~loc
-  | _ -> newvar ()
+      ignore @@
+      type_approx_constraint env (Pconstraint sty) ~loc ty_expected
+  | _ -> ()
 
-and type_approx_function env params c body ~loc =
-  (* We can approximate types up to the first newtype parameter, whereupon
-     we give up.
-  *)
-  match params with
-  | { pparam_desc = Pparam_val (label, default, pat) } :: params ->
-      type_approx_fun env label default pat
-        (type_approx_function env params c body ~loc)
-  | { pparam_desc = Pparam_newtype _ } :: _ ->
-      newvar ()
-  | [] ->
-    let body_ty =
-      match body with
-      | Pfunction_body body ->
-          type_approx env body
-      | Pfunction_cases ({pc_rhs = e} :: _, _, _) ->
-          newty (Tarrow (Nolabel, newvar (), type_approx env e, commu_ok))
-      | Pfunction_cases ([], _, _) ->
-          newvar ()
-    in
-    type_approx_constraint_opt env body_ty c ~loc
+and type_tuple_approx (env: Env.t) loc ty_expected l =
+  let labeled_tys = List.map (fun (label, _) -> label, newvar ()) l in
+  let ty = newty (Ttuple labeled_tys) in
+  begin try unify env ty ty_expected with Unify err ->
+    raise(Error(loc, env, Expr_type_clash (err, None, None)))
+  end;
+  List.iter2
+    (fun (_, e) (_, ty) -> type_approx env e ty)
+    l labeled_tys
+
+and type_approx_function =
+  let rec loop env params c body ty_expected ~in_function ~first =
+    let loc_function, _ = in_function in
+    let loc = loc_rest_of_function ~first ~loc_function params body in
+    (* We can approximate types up to the first newtype parameter, whereupon
+      we give up.
+    *)
+    match params with
+    | { pparam_desc = Pparam_val (label, default, pat) } :: params ->
+        let ty_res =
+          type_approx_fun_one_param env label default (Some pat) ty_expected
+            ~first ~in_function
+        in
+        loop env params c body ty_res ~in_function ~first:false
+    | { pparam_desc = Pparam_newtype _ } :: _ -> ()
+    | [] ->
+        (* In the [Pconstraint] case, we override the [ty_expected] that
+          gets passed to the approximating of the rest of the type.
+        *)
+        let ty_expected =
+          type_approx_constraint_opt env c ty_expected ~loc
+        in
+        match body with
+        | Pfunction_body body ->
+            type_approx env body ty_expected
+        | Pfunction_cases ({pc_rhs = e} :: _, _, _) ->
+            let ty_res =
+              type_approx_fun_one_param env Nolabel None None ty_expected
+                ~in_function ~first
+            in
+            type_approx env e ty_res
+        | Pfunction_cases ([], _, _) ->  ()
```

Comment:
This case in not reachable from the parser. I could be nice to add a comment explaining that here.
(the implementation is indeed what should be done if this case was reachable)